### DISCUSSION TITLE
[UI] Consistency as code for dialogs

### DISCUSSION THREAD

**Discussion by celian-garcia** [perses/perses#935] [Design Discussions]
This discussion is coming out of a PR that I recently created in the repository adding two new dialogs to Create and Delete perses projects. I struggled to find an example to inspire on, and felt like I had to either do duplication or bring some inconsistency.
https://github.com/perses/perses/pull/916#discussion_r1084397603

#### **IMHO, we could benefit thinking of a common component to handle the common parts of the dialogs.**

Probably starting with the deletion confirmation dialogs that are the ones with the most common wording and behavior.

## Deletion of a project (DeleteProjectDialog) 
one of the dialogs I had to create
```html
    <Dialog open={open} onClose={triggerOnClose}>
      <DialogTitle>Delete Project</DialogTitle>
      <DialogContent>
        Are you sure you want to delete Project {name}? This will delete all the dashboards within the project.
      </DialogContent>
      <DialogActions>
        <Button variant="contained" onClick={triggerOnConfirm}>
          Delete
        </Button>
        <Button variant="contained" onClick={triggerOnClose} color="secondary">
          Cancel
        </Button>
      </DialogActions>
    </Dialog>
```

## Deletion of a panel group
Already an icon button that I forgot to create in first place. Shall I duplicate it again with all this sx and so on?
Oh and I forgot the dividers...
```html
    <Dialog open={deletePanelGroupDialog !== undefined}>
      <DialogTitle>Delete Panel Group</DialogTitle>
      <IconButton
        aria-label="Close"
        onClick={() => closeDeletePanelGroupDialog()}
        sx={(theme) => ({
          position: 'absolute',
          top: theme.spacing(0.5),
          right: theme.spacing(0.5),
        })}
      >
        <CloseIcon />
      </IconButton>
      <form onSubmit={handleDelete}>
        <DialogContent dividers sx={{ width: '500px' }}>
          Are you sure you want to delete {deletePanelGroupDialog?.panelGroupName ?? 'panel group'}? This will delete
          all the panels within the group.
        </DialogContent>
        <DialogActions>
          <Button variant="contained" type="submit">
            Delete
          </Button>
          <Button variant="outlined" onClick={() => closeDeletePanelGroupDialog()}>
            Cancel
          </Button>
        </DialogActions>
      </form>
    </Dialog>
```

## Deletion of a panel
Sensitively the same display like the delete panel group, but with a different way to manage the deletion
```html
    <Dialog open={deletePanelDialog !== undefined}>
      <DialogTitle>Delete Panel</DialogTitle>
      <IconButton
        aria-label="Close"
        onClick={() => closeDeletePanelDialog()}
        sx={(theme) => ({
          position: 'absolute',
          top: theme.spacing(0.5),
          right: theme.spacing(0.5),
        })}
      >
        <CloseIcon />
      </IconButton>
      {deletePanelDialog && <DeletePanelForm deletePanelDialog={deletePanelDialog} />}
    </Dialog>
```

## Add/Edit panel group
An interesting try to already make one component for two things: creation and edition
```html
    <Dialog open={isOpen} TransitionProps={{ onExited: handleExited }}>
      {panelGroupEditor !== undefined && (
        <>
          <DialogTitle>{panelGroupEditor.mode} Panel Group</DialogTitle>
          <IconButton
            aria-label="Close"
            onClick={panelGroupEditor.close}
            sx={(theme) => ({
              position: 'absolute',
              top: theme.spacing(0.5),
              right: theme.spacing(0.5),
            })}
          >
            <CloseIcon />
          </IconButton>
          <DialogContent dividers sx={{ width: '500px' }}>
            <PanelGroupEditorForm initialValues={panelGroupEditor.initialValues} onSubmit={handleSubmit} />
          </DialogContent>
          <DialogActions>
            <Button variant="contained" type="submit" form={panelGroupEditorFormId}>
              {panelGroupEditor.mode === 'Edit' ? 'Apply' : 'Add'}
            </Button>
            <Button variant="outlined" onClick={panelGroupEditor.close}>
              Cancel
            </Button>
          </DialogActions>
        </>
      )}
    </Dialog>
```

**Comment by eunicorn** [comment]
Good write up, and i agree!

Some of the common designs I notice are -

1. Close icon button at the top right corner
2. Dividers for dialog content
3. For dialog actions, we always want primary button with `contained` variant on the left and secondary button (usually Cancel) with `outlined` variant on the right

Maybe we can start with these first when trying to create a generic `Dialog` component

**Reply by celian-garcia** [reply]
We are fully aligned on the common design 

Technically I'd need to try out but anything I can think of right now is something following the Dialog component interface like 
```html
<AppDialog>
  
  <!-- embed the close button after title -->
  <AppDialogTitle>
    {myTitle}
  </AppDialogTitle>
  
  <!-- embed the dividers -->
  <AppDialogContent>
    {mySentence} blabla or form
  </AppDialogContent>

  <!-- ensuring some button order logic based on primary vs secondary ? if it's possible :) -->
  <AppDialogActions>
      <!-- some new buttons -->
      <PrimaryButton onClick={handleSubmit}>Delete</PrimaryButton>
      <SecondaryButton  onClick={handleClose} >Cancel</SecondaryButton
  </AppDialogActions>
</AppDialog>
```

**Reply by celian-garcia** [reply]
Seeing the last example we could also benefit having an ActionButton + CancelButton component
```html
<ActionButton action={action} />
<CancelButton />
```

**Reply by eunicorn** [reply]
That looks good to me! One suggestion is to also make button labels and handlers props so we can reuse these dialogs not just for delete and cancel

**Comment by eunicorn** [comment]
Actually another idea - what if instead of making one generic Dialog component, we have wrappers for each part of the dialog?  For example, we can have a generic `DialogHeader` component that will include the title and close button, or a generic `DialogContent` component that include the dividers. This approach might be more flexible while still maintaining consistency in design

**Reply by cndonovan** [reply]
+1 to the list of details that represent the Perses dialog style, and +1 to making it into modular pieces 

Just to write down what I was talking about in the Perses meeting, a nice pattern I've seen where you only have to import one component looks like this:

```
import { Dialog } from 'perses/components'

<Dialog.Header> 
   ...
</Dialog.Header>

<Dialog.Content>
  ...
</Dialog.Content>

<Dialog.Actions> 
  <Dialog.PrimaryButton>...</Dialog.PrimaryButton>
</Dialog.Actions>
```

With `CloseButton`, you could (a) automatically include `CloseButton` in `Dialog.Header` or (b) make `Dialog.CloseButton`. Similarly, with `CancelButton`, you could (a) automatically include `CancelButton` in `Dialog.Actions` or (b) make `Dialog.CancelButton`. 

I lean toward (a) for both, but I think both approaches are OK.

**Reply by celian-garcia** [reply]
Sure! so we could use them independently if needed 👍  I'm curious to see the implementation of the ``DialogHeader`` comp. I'm wondering now if it's not easier to just have a ``CloseDialogButton`` component actually WDYT ?

**Edit**: We were writing at the same time. 😄 So we've got the same idea with this CloseButton. I'd tend to prefer b) as it sounds more modular, but I'm not expert enough in React to be sure it's possible.

**Reply by eunicorn** [reply]
I think we always want a close button in the header for every dialog so I don't think it's necessary to have a `Dialog.CloseButton` for now. If we ever have a use case in the future where we don't need the close button, then we can add it!

**Reply by eunicorn** [reply]
let me know what you all think - https://github.com/perses/perses/pull/948 :)

**Comment by celian-garcia** [comment]
Going through the React documentation. I found something that could be interesting. We can set in the theme the default value for ``dividers`` property of ``DialogContent``. https://mui.com/material-ui/customization/theme-components/#default-props

I'm wondering if we can't just rely on some similar mechanisms with other properties/styles that are always set to same values.

**Reply by eunicorn** [reply]
The issue we found with default props is that they don't get picked up when importing these components in other apps unless you also import the theme
